Implement Forgot Password Functionality

Ekep-Obasi opened the pull request:
Closes: #119 

## Summary

This PR implements Forgot Password Functionality

## Tests

- [x] Integration tests
- [ ] Unit Tests

## Additional information

This PR introduces a `useAsync` hook to make handling async operations more consistent across our components.

### **How `useAsync` works in `ForgotPasswordForm`**
I used the hook to manage password recovery requests. It handles the loading states, errors, and success callbacks automatically, so the component can focus on what it needs to do when things succeed or fail (like starting countdown timers or showing error messages). There's also a handy `reset` function to clear everything when needed.

This approach keeps the async stuff separate from the UI logic, which makes the code cleaner and easier to follow (my suggestion ?)

### **What we were doing before**
Previously, components like `SignInForm` were managing all their async operations directly inside the component. This meant we were writing the same loading state logic, error handling, and API calls over and over again in different places.

Ekep-Obasi committed:
feat(email): add  for password reset emails

- Implements email sending with reset URL and HTML content.

Ekep-Obasi committed:
feat(auth): add forgot password functionality

- Implemented  route to handle password recovery requests.
- Sends a password reset email with a secure token if the email exists.
- Added KV namespace for  to store reset tokens.
- Ensured responses do not reveal whether an email exists for security purposes.

Ekep-Obasi committed:
fix(docs): update Swagger UI URL to include  prefix

- Corrected the URL for  to point to  to resolve routing issues.

Ekep-Obasi committed:
test(integration): add tests for forgot password edge cases

Ekep-Obasi committed:
feat(hooks): add  hook for handling async operations

- Provides state management for data, loading, and error states.
- Supports options for enabling, success, and error callbacks.

Ekep-Obasi committed:
feat(ui): implement forgot password page and functionality

Ekep-Obasi committed:
chore(ui): centralize page layout and enhance header component customization

- Updated  to centralize the page layout for better alignment.
- Extended  and  components to allow more customization via props.

manvinderjit commented on src/components/Header/Header.tsx:
```diff
@@ -1,17 +1,18 @@
 
-const Header = ({ children, isLogoLarge = true }: Props) => (
+const Header = ({ children, isLogoLarge = true, textProps }: Props) => (
 	<header id='header-with-logo'>
 		<Logo isLarge={isLogoLarge} />
-		<HeaderTitle>{children}</HeaderTitle>
+		<HeaderTitle {...textProps}>{children}</HeaderTitle>
```
Out of curiosity: Is there an advantage or case where passing css styles as props is preferred instead of specifying the styles directly in `HeaderTitle.css`

Ekep-Obasi replied:
Since the component is used across most auth pages, I needed to make it more flexible for my use-case and couldn't think of a way to manage that just with css

kbventures replied:
@madcampos

madcampos replied:
@manvinderjit that depends. For super generic components like `Card` or `ColumnLayout`, it is a valid way to customize the component. If you are using css-in-js, it also makes sense as this is how the thing to do. But css-in-js is a terrible idea.

On this case as the `HeaderTitle` is very consistent and only have small changes. It doesn't make much sense to change it.

@Ekep-Obasi maybe it makes more sense to use another component here or confirming with the design team if it is okay to use the same component here?

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+.password-recovery-form {
+	border: 1px solid #cccccc;
```
suggestion:  There is a variable available in `src/csslayers/design-system.css` layer for border styles. Recommend using this for easy adjustment in future.
`73 --border: var(--border-base) solid var(--color-border);`

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+
+.input-block input {
+	background-color: var(--color-button);
+	border: solid 1px #cccccc;
```
Please use border variable here as well.

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+}
+
+.input-block label {
+	color: #333333;
```
suggestion: For colors, there are variables available in the `design-system.css`. Please refer to these. 

![image](https://github.com/user-attachments/assets/9697299f-4780-482d-8ae1-70f73e25e346)

manvinderjit replied:
If a color is unavailable, please add it here or ask the design team for suggestions.

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+
+hr {
+	border: none;
+	border-top: 1px solid #cccccc;
```
It would be better to use individual variables here. Please do the same. 
`var(--border-base) solid var(--color-border);`

manvinderjit commented on src/css-layers/pages.css:
```diff
@@ -1,14 +1,18 @@
+	height: inherit;
 	justify-content: center;
-	max-width: 400px;
+	width: clamp(20rem, 90vw, 25rem);
```
suggestion: `16.875rem` is the minimum I would recommend so that it covers devices with width as low as 320px, like older Androids and iPhones. Not sure on the max limit though as it varies by page.

Ekep-Obasi replied:
Agreed, thanks for the suggestion!

manvinderjit commented:
Praise: This is excellent work @Ekep-Obasi . Really well written and documented code.

Ekep-Obasi committed:
chore: replaced hardcoded css properties with design-system css variables

kbventures commented on src/css-layers/pages.css:
```diff
@@ -1,14 +1,18 @@
+#root { height: inherit; }
```
Question: can you explain rational behind this change that will affect all pages?

madcampos replied:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** ‏Why all the `inherit`ed `height`s?

A flex layout should already take all of the space available compared to flow layout. If you really really want to make the page take all the available space use this instead:

```css
html, body { height: 100%; }
body { margin: 0; }
```

All the flex containers should take all the available vertical space after that.

kbventures replied:
It results in this problem. 

<img width="429" height="704" alt="image" src="https://github.com/user-attachments/assets/2be9815b-7211-45db-8b51-0763bb70ce45" />

The TorontoJS banner is cut off.

kbventures commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -1,5 +1,5 @@
 # Sign in intergration test
```
There have been major changes made to how we do intergration testings.

https://github.com/torontojs/communityhub/pull/117/files

Please refer to this file and update your tests accordingly. 

There's no need to assert code in eveyr case you can simply write HTTP 200 in a lot of cases.

kbventures replied:
@Ekep-Obasi more changes are coming to intergration testing very soon.  We'll be introducing env variables instead of passing an options argument with terminal commands.  It might make sesnse to do do your intergration/unit testing in a a seperate pull request issue(only this time as we are still setting up Ci and unit/intergreation tests setups.  Once this ocify we can implement test for respective features we in the same PR).

kbventures commented on src/hooks/useAsync.ts:
```diff
@@ -0,0 +1,59 @@
+import { useCallback, useEffect, useState } from 'react';
```
I'm in favor of creating a hook, especially if it can be reused across pages. That said, we should also consider maintainability, loose coupling, ease of use, and testability.

Can we achieve creating a reusable adaptable hook that is easy to read, maintain and tests and can we dig deeper since this hook as why we NEED it considering it contains more code  than many pages in the front end so it probably deserves more explanation/rational  to support why it should exists.

madcampos commented on api/src/index.ts:
```diff
@@ -67,7 +67,7 @@ apiRoutes.doc('/open-api.json', {
 });
 
-apiRoutes.get('/docs', swaggerUI({ url: '/open-api.json' }));
+apiRoutes.get('/docs', swaggerUI({ url: '/api/open-api.json' }));
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
**praise:** Nice catch‏

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -221,6 +221,64 @@ authRoutes.openapi(
+	async (context) => {
+		const { email } = context.req.valid('json');
+
+		const response = { message: 'If an account with that e-mail exists, a password reset link has been sent.' };
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
**suggestion:** Let's add a comment here about not exposing existing emails and security. Like‏ the one we have on other auth endpoints.

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -221,6 +221,64 @@ authRoutes.openapi(
+			resetToken,
+			email,
+			{ expirationTtl: TEN_MINUTES_IN_SECONDS }
+		);
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Please move this code to `api/utils/‏auth.ts`, that is where the code that handles authentication lives. I think we even have some constants for timeout there.

madcampos commented on api/src/email/index.ts:
```diff
@@ -57,3 +57,32 @@ export async function sendAccountConfirmationEmail(context: Context, {
+	senderEmail: string;
+}
+
+export async function sendPasswordResetEmail(context: Context, {
+	token,
+	email,
+	apiKey,
+	senderEmail
+}: PasswordResetEmailParams) {
+	const resetUrl = new URL(`/pages/reset-password?token=${token}`, context.req.url).toString();
+
+	return sendEmail(context, {
+		apiKey,
+		from: senderEmail,
+		to: email,
+		subject: '[TorontoJS] Reset your password',
+		text: `You requested a password reset.\n\nPlease reset your password by visiting: ${resetUrl}`,
+		html: `
+			<h2>You requested a password reset</h2>
+			<p>Please reset your password by visiting the link below</p>
+			<p><a href="${resetUrl}">${resetUrl}</a></p>
+		`
+	});
+}
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** ‏You already have `context` as one of the parameters, the environment variables should be inferred from it here and not passed as parameters.

This provides better encapsulation for the consumer and don't expose implementation details. If we change from Resend to another email provider, we would only have to change this function implementation and not every call to it.

```suggestion
export async function sendPasswordResetEmail(context: Context, token: string, email: string) {
	const resetUrl = new URL(`/pages/reset-password?token=${token}`, context.req.url).toString();

	return sendEmail(context, {
		apiKey: context.env.RESEND_API_KEY,
		from: context.env.SENDER_EMAIL,
		to: email,
		subject: '[TorontoJS] Reset your password',
		text: `You requested a password reset.\n\nPlease reset your password by visiting: ${resetUrl}`,
		html: `
			<h2>You requested a password reset</h2>
			<p>Please reset your password by visiting the link below</p>
			<p><a href="${resetUrl}">${resetUrl}</a></p>
		`
	});
}
```

Ekep-Obasi replied:
@madcampos, this makes! I just followed the same structure of the `sendAccountConfirmationEmail` here https://github.com/torontojs/communityhub/blob/280ad415e6412abccd9a9541169ab65874cc84a0/api/src/email/index.ts#L39

I'll just modify both here.

madcampos replied:
Okay. This is probably an oversight there as well. Changing both would be great

madcampos commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -80,3 +80,63 @@ status == 404
+HTTP 200
+[Asserts]
+status == 200
+jsonpath "$.message" == "If an account with that e-mail exists, a password reset link has been sent."
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** I think hurl already does the assertion of the http status code from the `HTTP ...` line. Do we really need these?‏

Also, here just testing for the response is enough. Asserting on the message is brittle as if we change it in the future we would also have to fix the test. The message itself is not relevant, as it is a hint for consumers of the API, what really says if things or not in this case is the status code.

madcampos commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -80,3 +80,63 @@ status == 404
+HTTP 200
+[Asserts]
+status == 200
+jsonpath "$.message" == "If an account with that e-mail exists, a password reset link has been sent."
```
Same thing here.

madcampos commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -80,3 +80,63 @@ status == 404
+
+HTTP 422
+[Asserts]
+status == 422
```
Same here

madcampos commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -80,3 +80,63 @@ status == 404
+
+HTTP 422
+[Asserts]
+status == 422
```
Same here

madcampos commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -80,3 +80,63 @@ status == 404
+POST http://localhost:3000/api/auth/forgot-password
+Content-Type: application/json
+
+{}
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise.  -->
**praise:** Nice test‏

madcampos commented on api/src/routes/auth/integration-test.hurl:
```diff
@@ -80,3 +80,63 @@ status == 404
+
+HTTP 422
+[Asserts]
+status == 422
```
Same here

madcampos commented on api/src/routes/auth/validation.ts:
```diff
@@ -30,3 +30,11 @@ export type SignUpData = z.infer<typeof SignUpSchema>;
 export type ActivateData = z.infer<typeof ActivateSchema>;
+
+export const ForgotPasswordSchema = z.object({
+	email: z.string({ required_error: 'Email is required' })
+		.min(1, 'Email must be at least one character long')
+		.email('This is not a valid e-mail address')
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** [`email`](https://zod.dev/api?id=emails) already tests for a non empty string‏
```suggestion
	email: z.string({ required_error: 'Email is required' })
		.email('This is not a valid e-mail address')
```

Ekep-Obasi replied:
I followed this from other instances in the same file. I'll just correct them as well

madcampos commented on wrangler.toml:
```diff
@@ -68,6 +73,11 @@ binding = "ActivationTokens"
 
+[[env.dev.kv_namespaces]]
+binding = "PasswordResetTokens"
+id = "682d719e25304b57a5c95db6815da556"
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** Where does this id come from?‏

madcampos commented on wrangler.toml:
```diff
@@ -89,3 +99,8 @@ preview_id = "vms-production-session-tokens"
+
+[[env.production.kv_namespaces]]
+binding = "PasswordResetTokens"
+id = "f917eef00a5f4146a2d8d8fc2d274fba"
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**question:** Where does this id come from?‏

madcampos commented on src/hooks/useCountDown.ts:
```diff
@@ -0,0 +1,31 @@
+import { useCallback, useEffect, useState } from 'react';
+
+const ONE_MINUTE_IN_MS = 60000;
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏
```suggestion
const ONE_MINUTE_IN_MS = 60 * 1000;
```

madcampos commented on src/hooks/useCountDown.ts:
```diff
@@ -0,0 +1,31 @@
+
+const ONE_MINUTE_IN_MS = 60000;
+
+export const useCountdown = (initialMinutes = 0) => {
+	const [minutes, setMinutes] = useState(initialMinutes);
+	const [isActive, setIsActive] = useState(false);
+
+	useEffect(() => {
+		if (isActive && minutes > 0) {
+			const timer = setTimeout(() => {
+				setMinutes(minutes - 1);
+			}, ONE_MINUTE_IN_MS);
+			return () => clearTimeout(timer);
+		} else if (minutes === 0) {
+			setIsActive(false);
+		}
+		return () => undefined;
+	}, [minutes, isActive]);
+
+	const start = useCallback((mins: number = initialMinutes) => {
+		setMinutes(mins);
+		setIsActive(true);
+	}, [initialMinutes]);
+
+	return {
+		minutes,
+		isFinished: minutes === 0,
+		start
+	};
+};
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Use [`setInterval`](https://developer.mozilla.org/en-US/docs/Web/API/Window/setInterval) instead of `setTimeout`, it does wat you want[^1] with less logic, i.e. not needing to keep the `isActive` state.

Here is an example:
```typescript
export function useCountdown(initialTime) {
  const [timeRemaining, setTimeRemaining] = useState(initialTime);

  /**
   * Reset the timer and start a new countdown
   */
  const startCountdown = () => {
    setTimeRemaining(initialTime);

    if (timeRemaining > 0) {
      const interval = setInterval(() => {
        if (timeRemaining === 0) {
          clearInterval(interval);
          return;
        }
  
        setTimeRemaining(timeRemaining -1);
      }, ONE_MINUTE_IN_MS);
    }
  }

  return {
    timeRemaining,
    startCountdown
  };
}
```

Also a nitpick: prefer function declaration over arro functions whenever possible.

[^1]: One caveat to `setInterval`, and `setTimeout` for that matter, is they are not exactly precise, so there is drift in the time for both of them. But, for our use case they are good enough.

Ekep-Obasi replied:
@madcampos, I noticed the setInterval approach might have a cleanup issue. if startCountdown gets called multiple times or the component unmounts, we could end up with multiple running intervals. Would using useEffect with proper cleanup be safer here?

madcampos replied:
The `if` check on the `startCountdown` function will verify if the timer is already running. If you call this function multiple times and the timer is already running it will simply continue to run and won't start a new interval.

Looking from another side, the page is not intended to be long lived (i.e. running for hours or days on an open tab) so long tasks and memory leaks are not big issues here.

What we can do to improve the code is move the `startCountdown` outside of the hook so the function and all it's internal variable references will still be there in the closure.

And we could also move the variable binding for the interval outside of the hook, so the reference is always kept regardless. A module keeps the variable bindings so importing the same module again will not introduce new variables and instead will reference the same existing ones.

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+}
+
+.recovery-content p {
+	font-size: 1.125rem;
```
**question:** Don't we already have a variable for that?‏

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+.password-recovery-form {
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** As a general rule of thumb: if the element will be repeated multiple times, use a class, if it is going to be only one of it, use an id.‏

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+	gap: var(--spacing-md);
+}
+
+.success-message p {
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** This rule is a du[licate of another one, if you want to change the `<p>` elements in an entire block, target them in like this:‏
```
<grandparent element> p { ... }
```

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+
+.error-message::before {
+	color: var(--color-red);
+	content: '❌';
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** This content is not accessible and hard to style. Please prefer an inline svg instead.‏

Here is a place to find svg icons: https://icon-sets.iconify.design/

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+	border: var(--border);
+	border-radius: var(--rounded-xs);
+	box-sizing: border-box;
+	color: #333333;
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Please use variables for colors.‏

madcampos replied:
@kbventures please change this to a color variable.

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+}
+
+.input-block input:disabled {
+	background-color: #e5e5e5;
+	color: #999999;
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Please use variables for colors.‏

madcampos replied:
@kbventures here as well

madcampos commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,110 @@
+}
+
+.input-block input.error {
+	background-color: #fff5f5;
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Please use variables for colors.‏

madcampos replied:
@kbventures here as well

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+						onChange={handleEmailChange}
+						className={validationError ? 'error' : ''}
+						disabled={isSubmitted}
+						aria-invalid={!!validationError}
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** Do not use double negation for casting a value to boolean, use proper checks.‏

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+						isLarge
+						isPrimary
+						disabled={isLoading || isSubmitted || isEmailEmpty}
+						aria-describedby={isSubmitted ? 'countdown-status' : undefined}
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** ‏This may be confusing, no need to add it here as the timer is below on the page anyways.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+
+					{isSubmitted && (
+						<>
+							<div className='success-message' role='status' aria-live='polite'>
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Please remove the `role` and `aria-live` attributes they are excessive.‏

Adding it will make screen readers read it, but not necessarily in a sensible order, and because react destroys and recrates the dom, it will happen every re-render.

Just remove it as a user can read the page again. This is a simple enough page that skimming through the content of the page again should be fine.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+							<hr />
+							<div className='resend-timer' id='countdown-status' aria-live='polite'>
+								<ClockIcon aria-hidden='true' />
+								<p>Wait {minutes} {minutes === 1 ? 'minute' : 'minutes'} for a new activation link</p>
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Do not do internationalization by hand, use [`Intl.PluralRules`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules)‏ instead.

Ekep-Obasi replied:
@madcampos, this is so cool! thanks 🙌

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+								<p>Check your e-mail. We've sent you a recovery link. If you didn't get the link, check your spam folder or try again in 10 minutes.</p>
+							</div>
+							<hr />
+							<div className='resend-timer' id='countdown-status' aria-live='polite'>
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** Accessible timers are _very tricky_ to get right. I suggest for now to remove the `aria-live` attribute here and research more on how to do this properly.‏

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+		if (!emailInputRef.current?.checkValidity()) {
+			setValidationError('This is not a valid email address');
+			emailInputRef.current?.focus();
+			return;
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue:** The validation message here is redundant as the constraint validation api for forms already adds a message and displays it to the user.

Also, moving the focus unprompted can be a failure on [WCAG SC 3.2.1](https://www.w3.org/WAI/WCAG21/Understanding/on-focus.html).

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,139 @@
+						placeholder='Your account e-mail'
+						required
+						ref={emailInputRef}
+						onChange={handleEmailChange}
```
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought:** If we are not doing any validation with the email, there is no need to handle a change event, one less state for email. Just let html control the field normally.‏

madcampos commented on src/hooks/useAsync.ts:
```diff
@@ -0,0 +1,59 @@
+	reset(): void;
+}
+
+export function useAsync<T, Args extends readonly unknown[] = []>(
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** `T` is not a descriptive name, please pick a name that better describes what this generic is.‏

madcampos commented on src/hooks/useAsync.ts:
I haven't finished reviewing yet, but this code is very complex and convoluted. It seems over engineered to me.

Yes, it is interesting to have a helper for async code, but at the same time, we don't need that many bells and whistles. It needs to:
- Call a function
- Return the data from that function
- Return a potential error
- Update a loading state

In this code it seems like `enabled` and `isLoading` are doing the same thing. Also, we don't need calls to different situations and data types from the async operation, the consumer of this hook should handle and care about the data or error that resulted from the async call, not the hook.

madcampos commented on src/hooks/useAsync.ts:
```diff
@@ -0,0 +1,59 @@
+): UseAsyncState<T, Args> {
+	const { enabled = false, onSuccess, onError } = options;
+
+	const [data, setData] = useState<T | null>(null);
+	const [error, setError] = useState<Error | null>(null);
+	const [isLoading, setIsLoading] = useState(false);
+
+	const execute = useCallback(async (...args: Args): Promise<void> => {
+		if (!asyncFn) { return; }
+
+		try {
+			setIsLoading(true);
+			setError(null);
+
+			const result = await asyncFn(...args);
+			setData(result);
+			onSuccess?.(result);
+		} catch (err) {
+			const resolvedError = err instanceof Error ? err : new Error('Something went wrong');
+			setError(resolvedError);
+			onError?.(resolvedError);
+		} finally {
+			setIsLoading(false);
+		}
+	}, [asyncFn, onSuccess, onError]);
+
+	const reset = useCallback((): void => {
+		setData(null);
+		setError(null);
+		setIsLoading(false);
+	}, []);
+
+	useEffect(() => {
+		if (enabled && asyncFn) {
+			void (execute as () => Promise<void>)();
+		}
+	}, [enabled, execute]);
+
+	return { data, error, isLoading, execute, reset };
+}
```
```suggestion
export function useAsync<Data>(asyncFunction: Function) {
	const [data, setData] = useState<Data | undefined>(undefined);
	const [error, setError] = useState<Error | undefined>(undefined);
	const [isLoading, setIsLoading] = useState(false);

	const execute = async (...args: unknown[]) => {
		try {
			setIsLoading(true);
			setError(undefined);

			const result = await asyncFunction(...args);
			setData(result);
		} catch (err) {
			setError(err);
		} finally {
			setIsLoading(false);
		}
	};

	return { data, error, isLoading, execute };
}
```

Some notes here:
- React hooks will memoize the results, track dependencies and all that. We don't care about it here, the burden of executing a call again should be on the component, so a simple async function will do. The component consuming the code needs to check if they don't have data, nor error and then execute the function.
- Please use `undefined` over `null`, they have subtle differences. The main one being `null instanceof Object === true`.
- Do not overtype things, the compiler is your friend, let it do it's job and infer most of the types for you. Only add types if the compiler cannot figure that out otherwise or you are absolutely certain the thing is what it is expected to be.
- `unknown` can fit any type and is better than `any` (in fact eslint complains about using `any`), it can accept anything and is useful when we don't care about the type of things.
- `await`ing on a regular function will just return the result of that function, it won't cause a problem. Also, deeply nested promises get unwrapped when they are `await`ed.

manvinderjit committed:
feat: add HTML and TS interfaces for TextInputComponent

manvinderjit committed:
feat: update props, add labelSlot

manvinderjit committed:
feat: update HTML for disabled, readonly

manvinderjit committed:
feat: add css styles for text input

manvinderjit committed:
fix: move disabled text input handler to parent

manvinderjit committed:
fix: fix heading tag hierarchy

manvinderjit committed:
refactor: simplify props and add disabled button handler

simplify props to use built in props
add a disabled button handler within the component

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: add imports and handler

manvinderjit committed:
fix: update sample usage props

manvinderjit committed:
fix: updates sample inputs to use default disabled handler

manvinderjit committed:
fix: update css selector for disabled input labels

manvinderjit committed:
cleanup: remove unused code in sample text inputs

manvinderjit committed:
fix: update css for nesting

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
Revert "fix: update css for nesting"

This reverts commit 79dce2ab18f25220a94fe7f0a6fc41b0d0486f32.

manvinderjit committed:
fix: update css to use layers

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: replace tags for appropriate usage

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
feat: add helperMessage slot to text input element

manvinderjit committed:
styles: switch inline css with designsytem, layers

switch inline css of sample text inputs to use the design system and layers.

kbventures committed:
Merge dev

kbventures committed:
Fix: wrangler

kbventures committed:
Fix: remove css

kbventures committed:
Fix: const number standardization

kbventures committed:
Update src/hooks/useAsync.ts

Co-authored-by: Marco Campos <me@madcampos.dev>

kbventures committed:
Update api/src/routes/auth/validation.ts

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: resolve formatting issues

kbventures committed:
Fix: Move forgot password to tests locaction and remove stale code

kbventures committed:
Reove unused hooks

kbventures committed:
Draft refactor

kbventures committed:
Valid email enable send link for forgot password

kbventures committed:
Fix: react attribute naming concention

kbventures committed:
Fix: heigh inherits at top lavel page unintentially hiding logo

kbventures committed:
One time usage should be id's not classes

kbventures committed:
Fix: pages back to 400px

kbventures committed:
Fix: use design system variables in forgot password css file

kbventures committed:
Forgot password page remove unused css

bcha92 commented on api/src/routes/auth/index.ts:
```diff
@@ -222,6 +222,64 @@ authRoutes.openapi(
+
+		const emailExists = await checkExistingEmail(context.env.Database, email);
+		if (!emailExists) {
+			return context.json(response satisfies StatusResponse, StatusCodes.OKAY);
```
Should this `StatusCodes.OKAY` response be changed to `StatusCodes.BAD_REQUEST` instead since this conditional looks for a falsy `emailExists` value?

Also do we have a default catcher for `StatusCodes.INTERNAL_SERVER_ERROR` or is this initiated by default?

kbventures replied:
@bcha92 This is because we're trying to obfuscate the existence of emails to stop someone from phishing for information. 

We do not, to the best of my knowledge, have a default catcher for 500 errors.  The only 500 status code caught in our API Tests is from an explicit try-catch instance.

bcha92 reviewed (commented):
Other than the one comment about StatusCodes, I think it looks okay.

kbventures committed:
Merge branch 'dev' into forgot-password

kbventures committed:
Removed unhandled 500 status code from forgot password route

kbventures committed:
Forgot password route checkint to see if an attempt to change password ttl is still alive before processing one

madcampos commented on src/hooks/useCountDown.ts:
```diff
@@ -0,0 +1,18 @@
+import { useEffect } from 'react';
+
+const MS = 1000;
+
+export function useCountdown(
+	seconds: number,
+	setSeconds: (s: number) => void
+) {
+	useEffect(() => {
+		if (seconds <= 0) { return; }
+		const timer = setTimeout(() => {
+			setSeconds(seconds - 1);
+		}, MS);
+		return () => clearTimeout(timer);
+	}, [seconds, setSeconds]);
+
+	return { isFinished: seconds <= 0 };
+}
```
```suggestion
import { useEffect, useState } from 'react';

const ONE_SECOND_IN_MS = 1000;

export function useCountdown(seconds: number) {
	const [remaningTime, setRemainingTime] = useState(seconds);

	useEffect(() => {
		const interval = setInterval(() => {
			setRemainingTime(remainingTime - 1);

			if (remainingTime <= 0) {
				clearInterval(interval);
			}
		}, seconds * ONE_SECOND_IN_MS);
	}, [seconds]);

	return { isFinished: remainingTime <= 0, remainingTime };
}
```

- Use `setInterval` instead of `setTimeout`, this is what the function is for.
- Keep the state _inside_ the hook.
- Return the remaining time, it is data from the hook and not external to it.

madcampos commented on wrangler.toml:
```diff
@@ -51,6 +51,11 @@ binding = "ActivationTokens"
 id = "vms-local-activation-token"
 preview_id = "vms-local-activation-token"
 
+[[kv_namespaces]]
+binding = "PasswordResetToken"
+id = "vms-local-password-reset-token"
+preview_id = "vms-local-password-reset-token"
+
```
**note**: this needs to be repeated for the other envs.

Use this:
```
[[env.dev.kv_namespaces]]
binding = "PasswordResetToken"
id = "5392322d41c44203bce89425fcc12f12"
preview_id = "vms-dev-password-reset-token"

[[env.prod.kv_namespaces]]
binding = "PasswordResetToken"
id = "08ad1ddf3a134007bbed3b9266cd6d1f"
preview_id = "vms-production-password-reset-token"
```

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -225,6 +225,72 @@ authRoutes.openapi(
+			return context.json(response satisfies StatusResponse, StatusCodes.OKAY);
+		}
+
+		// TODO: Discussion around wether we should create a access entry to deal with this state
```
**question**: Haven't we already discussed that? Is it okay to remove this comment?

madcampos replied:
@kbventures remove this comment please

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -225,6 +225,72 @@ authRoutes.openapi(
+		}
+
+		// TODO: Discussion around wether we should create a access entry to deal with this state
+		const ForgotPasswordList = await context.env.PasswordResetToken.list();
+
+		for (const key of ForgotPasswordList.keys) {
+			const keyName = await context.env.PasswordResetToken.get(key.name);
+			if (keyName === email) {
+				return context.json(response satisfies StatusResponse, StatusCodes.OKAY);
+			}
+		}
```
**issue**: do a get here and check for a null value instead

kbventures replied:
????

madcampos commented on api/src/routes/auth/index.ts:
**question**: When a user resets their password, to which endpoint they send the new password?

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,115 @@
+import Button from '../Button/Button.tsx';
+import ClockIcon from '../Icons/ClockIcon.tsx';
+
+// TODO: Investigate if we're able to use heartbeat to check if forgot password request already in progress and how the page react if there is one
```
To answer that: yes, we can add a check to the profile status for the password reset token, if it exists, we return that the user is in the "reset password" state.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,115 @@
+							<hr />
+							<div className='resend-timer' id='countdown-status'>
+								<ClockIcon aria-hidden='true' />
+								<p>Wait {Math.ceil(seconds / 60)} minutes for a new activation link</p>
```
```suggestion
								<p>Wait {Math.ceil(seconds / 60)} minutes to send a new activation link</p>
```

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,115 @@
+				setSeconds(600);
+			}
+		} catch (error) {
+			if (import.meta.env.MODE === 'development') {
+				if (error instanceof Error) {
+					console.error(error.name);
+					console.error(error.cause);
+					console.error(error.message);
+					console.error(error.stack);
+				} else {
+					throw new Error('Invalid email error');
+				}
+			}
+		}
```
**nitpick**: please cleanup this error log.

kbventures replied:
@madcampos This pattern is utilized throughout the project.  Please be more specific.

madcampos replied:
The `console` API allows objects, so the first part should only be:
```ts
console.log(error);
```

Also, re-throwing an error in a `catch` block seems like a code smell. Why are we `catch`ing the error if we are throwing it again? A `catch` block should _handle_ the error there

kbventures replied:
The purpose is to catch errors that aren’t instances of the Error class.

That said, in retrospect, this is overkill for something intended to simplify development and is disabled in production.

Changed all of them.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,115 @@
+		setButtonDisabled(true);
+		setEmailDisabled(true);
+
+		requestPasswordRecovery(emailValue);
```
**nitpick**: this function is only called here, please move the code here and remove the extra function

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,115 @@
+	const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
+		event.preventDefault();
+
+		const formData = new FormData(event.currentTarget);
```
```suggestion
		const formData = new FormData(event.target as HTMLFormElement);
```

Submit events don't bubble, this should work just fine.

kbventures replied:
Utilizing .target instead of .currentTarget triggers a TypeScript check error. 

```
Argument of type 'EventTarget' is not assignable to parameter of type 'HTMLFormElement'.
  Type 'EventTarget' is missing the following properties from type 'HTMLFormElement': acceptCharset, action, autocomplete, elements, and 319 more.ts(2345)
```

madcampos replied:
This is a shortcoming of typescript and event handlers. It is very annoying. You should cast to the element in this case.

I've updated the previous comment with the casting.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,115 @@
+		event.preventDefault();
+
+		const formData = new FormData(event.currentTarget);
+		const emailValue = (formData.get('email') as string).trim().toLowerCase();
```
I think this check should be moved to the backend, everywhere we have emails, we should normalize it to lowercase.

kbventures replied:
https://github.com/torontojs/communityhub/pull/199

madcampos commented on api/api_tests/19_auth_forgot_password.hurl:
The file name will conflict with another existing one. Why don't we move the tests to different folders, like this:
- routes
  - route a
    - api-tests
    - ...
  - route b
    - api-test
    - ...

manvinderjit committed:
feat: add useProfileRedirect hook

add profileRedirect hook that redirects user based on profile status.
update Complete Profile page to use hook.

madcampos committed:
feat: add trimming and converting emails to lower case

madcampos committed:
fix: extra spaces test

madcampos committed:
feat: add test for case insensitive email

manvinderjit committed:
fix: update status switch cases

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: remove unused redirect paths

madcampos committed:
fix: zod validation order

Co-authored-by: Manvinderjit Rupal <91133409+manvinderjit@users.noreply.github.com>

madcampos committed:
fix: zod validation order

kbventures committed:
Move create password reset fucntion to utils folder

kbventures committed:
Update src/hooks/useCountDown.ts

Co-authored-by: Marco Campos <me@madcampos.dev>

kbventures committed:
Update src/components/ForgotPassword/ForgotPassword.tsx

Co-authored-by: Marco Campos <me@madcampos.dev>

kbventures committed:
Update src/components/ForgotPassword/ForgotPassword.tsx

Co-authored-by: Marco Campos <me@madcampos.dev>

kbventures committed:
Add dependency to timer

kbventures committed:
Fix: setInterval replace setTimeout

kbventures committed:
Wrangler.toml local dev prod update for PasswordReset kv

kbventures committed:
Refactor: minimize development console log

kbventures committed:
Refactor: console.log for evelopment

kbventures committed:
Fix: utilizing event target instead of currentTarget for form

kbventures committed:
Reset Password Page HTML

kbventures committed:
Reset Password Vite Config, index.html and Component

kbventures committed:
Valid Component FE Forgot Password and updating email schema to match other schemas

kbventures committed:
Merg dev

kbventures committed:
Fix; duplicate zod requirement

kbventures committed:
Reset password page in main index.html file

kbventures committed:
Fix: design tweaks to align with implemented pages for the input reusable components

kbventures committed:
Reset password page

kbventures committed:
Handling passsword reset succss and failures

kbventures committed:
Reset password route and forgot password creation and refactor respectively

kbventures committed:
Reset Password Database Function

madcampos commented on api/src/routes/auth/data.ts:
```diff
@@ -162,3 +162,11 @@ export async function activateProfile(database: D1Database, email: string) {
+
+export async function resetPassword(database: D1Database, password: string, email: string) {
+	const { success } = await database
+		.prepare(`UPDATE ${DBTables.PROFILE} SET password = ? WHERE email ?`)
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Should we check as well if the profile is activated and not deleted?

madcampos commented on api/src/routes/auth/validation.ts:
```diff
@@ -78,3 +78,20 @@ export type SignUpData = z.infer<typeof SignUpSchema>;
+
+export const ResetPasswordSchema = z.object({
+	password: z.string({ error: 'Password is required' })
+		.min(15),
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Do not use a magic number. Without context, a plain number doesn't have much meaning.

Once the text fixes PR is merged, you can export this constant and reference it: https://github.com/torontojs/communityhub/blob/ed929ca6edef011324f9de9410db9c74876afc0a/api/src/utils/passwordStrengthCheck.ts#L4

madcampos commented on src/components/ResetPassword/ResetPassword.css:
```diff
@@ -0,0 +1,127 @@
+}
+
+.suggestion-icon {
+	background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none'%3E%3Cpath d='M19 3H5C3.89543 3 3 3.89543 3 5V19C3 20.1046 3.89543 21 5 21H19C20.1046 21 21 20.1046 21 19V5C21 3.89543 20.1046 3 19 3Z' stroke='black' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'/%3E%3Cpath d='M12 8V16' stroke='black' stroke ...
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Same as this comment https://github.com/torontojs/communityhub/pull/173#discussion_r2217986531

Also this comment on how to solve implement it: https://github.com/torontojs/communityhub/pull/173#discussion_r2223609295

kbventures replied:
https://github.com/torontojs/communityhub/issues/213

madcampos commented on src/components/ResetPassword/ResetPassword.css:
```diff
@@ -0,0 +1,127 @@
+	width: 1.5rem;
+
+	&.error-icon {
+		background-image: url("data:image/svg+xml,%3Csvg id='Layer_1' data-name='Layer 1' xmlns='http://www.w3.org/2000/svg' viewBox='0 0 122.88 111.54'%3E%3Cdefs%3E%3Cstyle%3E.cls-1%7Bfill:%23cf1f25;%7D.cls-2%7Bfill:%23fec901;fill-rule:evenodd;%7D.cls-3%7Bfill:%23010101;%7D%3C/style%3E%3C/defs%3E%3Ctitle%3Ewarning%3C/title%3E%3Cpath class='cls-1' d='M2.35,84.42,45.28,10.2l.17-.27h0A23,23,0,0,1,52.5,2. ...
```
Same here as the previous comment

madcampos commented on src/components/TextInputComponent/TextInputComponent.css:
```diff
@@ -13,18 +13,19 @@
 			border-radius: var(--rounded-xs);
 			gap: var(--spacing-xs);
-			height: auto;
+			height: 42px;
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Do not use pixel values. Furthermore, it is an input, which by default has a `display` of `inline`. Inline elements don't compute `height` and `width` like `block` elements. So this won't make much of a difference and may actually be prejudicial.

madcampos commented on src/components/TextInputComponent/TextInputComponent.css:
```diff
@@ -49,3 +50,30 @@
+	background-color: #e5e5e5;
+	color: var(--color-gray-500);
+	cursor: not-allowed;
+} */
```
**nitpick**: Please clean up commented code, we want to keep the repo tidy. Also, there seems to be a typo in the first two lines.

madcampos commented on src/index.html:
```diff
@@ -10,6 +10,8 @@
 		<ul>
 			<li><a href="/pages/sign-in/">Sign-In Page</a></li>
+			<li><a href="/pages/forgot-password/">Forgot Password Page</a></li>
+			<li><a href="/pages/reset-password/?token=cb81a6dc-7328-4e7a-b0a1-14e83ccb3468">Reset Password Page</a></li>
```
**question**: how will this token work? Doesn't the page check if the token is valid when loading the page?

kbventures replied:
@madcampos Marco, this branch is still marked "draft" :).  I needed a dummy UUID  to test the page while building it.

madcampos replied:
When testing "dummy UUIDs" please use one that is all the same character, so all 0s, all 1s, or something like that.

madcampos commented on src/pages/reset-password/index.html:
```diff
@@ -0,0 +1,12 @@
+	<head>
+		<meta charset="UTF-8" />
+		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
+		<title>Sign-In</title>
```
**issue**: the title here is wrong

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,141 @@
+		window.location.href = '/pages/sign-in';
+	}
+
+	// TODO how to handle errors from server
```
**question**: what kinds of errors?

We can simply show a message to users when there is an error.

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,141 @@
+	};
+
+	const resetPassword = async (token: string | null, password: string) => {
+		// setInvalidPasswordMessage('');
```
**nitpick**: please remove unused code to keep the repo tidy.

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,141 @@
+		} catch (error) {
+			if (import.meta.env.MODE === 'development') {
+				console.log(error);
+			}
```
**thought**: I really don't like this pattern of error reporting.
We should just use `console.error` (not `log`) to log the messages to the console anyways, regardless of environment...

This helps with debugging for tech savy users.

For non tech users we should provide a modal dialog or something with an error message, a reference code, and some UI that the users can use to help report the error.

kbventures replied:
@madcampos 

https://github.com/torontojs/communityhub/issues/210

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,141 @@
+							<div id='password-input-strength' aria-live='polite'>
+								<span>Password strength: {strengthLabels[strength || 0]}</span>
+							</div>
+							<div className='password-meter' data-password-strength={strengthLabels[strength || 0]} aria-hidden='true'>
```
**question**: why are we hiding the password meter?

madcampos replied:
The best markup for a password meter would be:
```jsx
<label for="password-meter">Password trength</label>
<div id="password-meter-wrapper" data-password-strength={strengthLabels[strength || 0]}>
  <meter
    className="sr-only"
    id="password-meter"
    min="0"
    {/* This is whatever the max score for the password is. */}
    max="4"
    value={passwordStength}
    {/* This is the minimum score for a bad password, in this case 0. */}
    low="0"
    {/* This is the minimum score for a "decent" password. I think for us it is 3. */}
    high="3"
    {/* This is the best value for a password. In our case this is the same as the decent one. */}
    optimum="3"
  />
  <div className='password-meter-level'></div>
  <div className='password-meter-level'></div>
  <div className='password-meter-level'></div>
</div>
```

This exposes the password meter as an actual meter, associates a label to it so it will actually be read as something that makes sense.

The 3 lines to show the password strength are shown just as squares without any extra semantics. We have to have them as styling the `<meter>` element is a nightmare.

We apply the `sr-only` class to hide the meter element so it doesn't show for all users. The divs will display as a visual equivalent.

kbventures replied:
> **question**: why are we hiding the password meter?

That's the paradigm utilized previously in other pages and still being requested through the current design(Figma).

kbventures replied:
https://github.com/torontojs/communityhub/issues/212

kbventures replied:
https://github.com/torontojs/communityhub/issues/212

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -234,6 +234,129 @@ authRoutes.openapi(
+
+		const emailExists = await checkExistingEmail(context.env.Database, email);
+
+		// We are deliberately sending OK statuscodes to disable an attackers ability to fish for information on exisiting emails
```
```suggestion
		// INFO: Hide non existing emails to reduce attack surface from guessing registered emails.
```

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -234,6 +234,129 @@ authRoutes.openapi(
+
+		const success = await resetPassword(context.env.Database, email, hashedPasswordWithSalt);
+
+		// TODO: Send email to user stating their password has been succesfully changed(Bonus: remedy if they didn't make the change)
```
Please resolve this TODO

kbventures committed:
Merge branch 'dev' into forgot-password

kbventures committed:
Auth Forgot Password Remove comment

kbventures committed:
Changed all console.log to console.error

kbventures committed:
Remove unsused commments

kbventures committed:
Fix: Reset Password Name

kbventures committed:
Fix: Removed unused notes from reset password page

kbventures committed:
fix: removed unused height in in put component

kbventures committed:
Remove notes from reset password page

bcha92 reviewed (dismissed):
Looks good so far~

kbventures committed:
Merge Dev

kbventures committed:
resetPassword function

kbventures committed:
Fixes

kbventures committed:
Refactor: removed unused hook async

kbventures committed:
ResetPassword page validating token before loading page

kbventures committed:
tautsesetPassword page validating token before loading page

kbventures committed:
Fix: react attributtes name fix and now sending email to next page from sign up page

kbventures committed:
Fix: added link to sign-up page from check your email page when incoret password utilized to sign up

bcha92 reviewed (dismissed):
I can't think of anything else to critique aside creating unit tests for the utility function `sendPasswordResetEmail`, but I think this is something we can put in the burner for now.

madcampos committed:
chore: add meeting notes from design meeting

manvinderjit commented on api/src/routes/auth/data.ts:
```diff
@@ -162,3 +162,11 @@ export async function activateProfile(database: D1Database, email: string) {
+
+export async function resetPassword(database: D1Database, password: string, email: string) {
+	const { success } = await database
+		.prepare(`UPDATE ${DBTables.ACCESS} SET password = ? WHERE email ? AND activatedAt IS NOT NULL AND deletedA IS NULL LIMIT 1`)
```
typo:
```suggestion
		.prepare(`UPDATE ${DBTables.ACCESS} SET password = ? WHERE email = ? AND activatedAt IS NOT NULL AND deletedAt IS NULL LIMIT 1`)
```

manvinderjit commented on api/src/routes/auth/index.ts:
```diff
@@ -234,6 +234,165 @@ authRoutes.openapi(
+
+		const hashedPasswordWithSalt = await hashPassword(password);
+
+		const success = await resetPassword(context.env.Database, email, hashedPasswordWithSalt);
```
typo: the order of arguments is incorrect

<img width="684" height="47" alt="image" src="https://github.com/user-attachments/assets/c53de3b9-9b16-4b6c-88a4-c66cf6b30caf" />

manvinderjit commented on api/src/routes/auth/index.ts:
```diff
@@ -234,6 +234,165 @@ authRoutes.openapi(
+			email
+		});
+
+		await context.env.PasswordResetToken.put(resetToken, email);
```
query: do we have a TTL for the token?

manvinderjit commented on api/src/routes/documents/api-tests/get-documents/19_auth_forgot_password.hurl:
```diff
@@ -0,0 +1,58 @@
+
+{
+  "email": "king.arthur@camelot.uk",
+  "password":"H0lyGr@il42!"
```
query: why is a user sending a password when they have forgotten their password?

kbventures replied:
Great question. This is @madcampos code.  :)

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,96 @@
+	border-radius: var(--rounded-md);
+	display: flex;
+	flex-direction: column;
+	max-width: 400px;
```
suggest: let us use rem for consistency
```suggestion
	max-width: 25rem;
```

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.css:
```diff
@@ -0,0 +1,96 @@
+	color: var(--color-gray-900);
+	font-size: 1rem;
+	font-weight: var(--weight-regular);
+	height: 42px;
```
suggest:
```suggestion
	height: 2.625rem;
```

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,116 @@
+
+		if (remainingTime === 0) { return; }
+
+		const timer = setInterval(() => {
+			setRemainingTime((prev) => prev - 1);
+		}, 1000);
+
+		return () => clearTimeout(timer);
```
query: why are we using `setInterval` and `clearTimeout` ? Shouldn't it be either setTimeout + clearTimeout or setInterval + clearInterval ?

madcampos replied:
I've commented that before... [TWO](https://github.com/torontojs/communityhub/pull/126#discussion_r2217075335) [TIMES](https://github.com/torontojs/communityhub/pull/126#discussion_r2144191119)!

madcampos replied:
It should be set/clear interval for both.

manvinderjit commented on src/components/ResetPassword/ResetPassword.css:
```diff
@@ -0,0 +1,127 @@
+.input-block label {
+	display: flex;
+	font-weight: var(--weight-semibold);
+	height: 29px;
```
suggest: rem > px

manvinderjit commented on src/components/TextInputComponent/TextInputComponent.css:
```diff
@@ -13,18 +13,18 @@
 		& > input {
-			background-color: var(--color-card);
+			font-size: var(--text-xs);
+			background-color: var(--color-button);
```
query: why are we using button color for input background?

manvinderjit commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,161 @@
+
+	useEffect(() => {
+		try {
+			const result = fetch('/api/auth/valid-reset-pw-token', {
```
typo: missing await

madcampos replied:
This should not be passing linting...

kbventures replied:
@madcampos 

https://github.com/torontojs/communityhub/issues/218

madcampos commented on api/src/routes/documents/api-tests/get-documents/19_auth_forgot_password.hurl:
**issue**: this file is in the wrong folder.

It should be under:
```
api/src/auth/api-tests/forgot-password/<num>-<test-name>.hurl
```

madcampos commented on api/src/routes/documents/api-tests/get-documents/19_auth_forgot_password.hurl:
```diff
@@ -0,0 +1,58 @@
+}
+
+HTTP 200
+[Asserts]
+status == 200
```
```suggestion
```
The previous line already asserts the status code.

madcampos commented on api/src/routes/documents/api-tests/get-documents/19_auth_forgot_password.hurl:
**issue**: also please break down different scenarios in different files. See other test files for reference.

madcampos commented on api/src/routes/documents/api-tests/get-documents/19_auth_forgot_password.hurl:
```diff
@@ -0,0 +1,58 @@
+# Forgot password with valid email
+POST http://localhost:3000/api/auth/forgot-password
+Content-Type: application/json
```
```suggestion
```

The content type is not required, from hurl docs: https://hurl.dev/docs/request.html#json-body

madcampos commented on api/src/routes/documents/api-tests/get-documents/19_auth_forgot_password.hurl:
```diff
@@ -0,0 +1,58 @@
+Content-Type: application/json
+
+{
+  "email": "king.arthur@camelot.uk",
+  "password":"H0lyGr@il42!"
```
**issue**: Please use variables for this. Preferably, get a new user that is not referenced in any of the previous tests just for this case to avoid conflicts.

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,161 @@
+		const formData = new FormData(event.target as HTMLFormElement);
+		const passwordValue = formData.get('password') as string;
+
+		setButtonDisabled(true);
+		setPasswordDisabled(true);
```
**question**: why are there two state variables to disable the form> This should only be one state variable.

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,161 @@
+		setButtonDisabled(true);
+		setPasswordDisabled(true);
+
+		await resetPassword(token, passwordValue);
```
**issue**: unless there is a redirect, reenable the form after the request is sent.

madcampos commented on src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,161 @@
+		await resetPassword(token, passwordValue);
+	};
+
+	const resetPassword = async (token: string | null, password: string) => {
+		try {
+			const response = await fetch('/api/auth/reset-password', {
+				method: 'POST',
+				headers: {
+					'Content-Type': 'application/json'
+				},
+				body: JSON.stringify({ token, password })
+			});
+			if (!response.ok) {
+				window.location.href = '/pages/sign-in';
+			} else {
+				window.location.href = '/pages/sign-in';
+			}
+		} catch (error) {
+			if (import.meta.env.MODE === 'development') {
+				console.error(error);
+			}
+		}
+	};
```
**issue**: If you are not using any state, move the function outside react's rendering function.

Big functions take longer to parse, execute, and process. Depending on the function size, it will bail out of compilation entirely and be interpreted making the code even slower.

On top of that, react is already slow, adding too many things inside it's rendering loop will make it even slower.

Also, use function declarations instead of function expressions whenever possible. It is in our coding style guidelines.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,116 @@
+		return () => clearTimeout(timer);
+	}, [remainingTime, isSubmitted]);
+
+	const handleValidEmail = () => {
+		setIsValid(!emailInputRef.current?.checkValidity());
+		if (validEmail) {
+			setButtonDisabled(true);
+		} else {
+			setButtonDisabled(false);
+		}
+	};
+
+	const requestPasswordRecovery = async (email: string) => {
+		try {
+			const response = await fetch('/api/auth/forgot-password', {
+				method: 'POST',
+				headers: { 'Content-Type': 'application/json' },
+				body: JSON.stringify({ email })
+			});
+			if (response.ok) {
+				setIsSubmitted(true);
+				setRemainingTime(600);
+			}
+		} catch (error) {
+			if (import.meta.env.MODE === 'development') {
+				console.error(error);
+			}
+		}
+	};
```
**issue**: move the logic outside of the main rendering function.
This can be refactored to simplify the need of so many states.

madcampos commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,116 @@
+
+const ForgotPasswordForm = (): React.JSX.Element => {
+	const emailInputRef = useRef<HTMLInputElement>(null);
+	const [isSubmitted, setIsSubmitted] = useState(false);
+	const [buttonDisabled, setButtonDisabled] = useState(true);
+	const [validEmail, setIsValid] = useState(false);
+	const [emailDisabled, setEmailDisabled] = useState(false);
+	const [remainingTime, setRemainingTime] = useState(0);
```
**issue**: There are too many overlapping states. Please think which ones of those can be combined together or derived from other variables?

kbventures replied:
I went with a derived state for simplicity.

madcampos commented on package.json:
```diff
@@ -37,7 +37,7 @@
 		"test:unit": "vitest run",
-		"test:api-single": "hurl --continue-on-error --jobs 4 --max-time 300 --test --variables-file .dev.vars --variables-file .api-test.vars",
+		"test:api-file": "hurl --continue-on-error --jobs 4 --max-time 300 --test --variables-file .dev.vars --variables-file .api-test.vars",
 		"test:api": "npm run test:api-single -- --glob \"**/*.hurl\"",
```
```suggestion
		"test:api": "npm run test:api-file -- --glob \"**/*.hurl\"",
```

ProTip: when renaming things, use search and replace as much as possible, do global searches and use regex searches.

kbventures committed:
Update api/src/routes/auth/data.ts

Co-authored-by: Manvinderjit Rupal <91133409+manvinderjit@users.noreply.github.com>

madcampos commented on api/src/routes/auth/validation.ts:
```diff
@@ -78,3 +79,26 @@ export type SignUpData = z.infer<typeof SignUpSchema>;
 export type ActivateData = z.infer<typeof ActivateSchema>;
+
+export const ForgotPasswordSchema = z.object({
+	email: z.string({ error: 'Email is required' })
+		.trim()
+		.toLowerCase()
+		.min(1)
```
**issue**: instead of creating new schemas reference existing ones.

```suggestion
	email: ProfileSchema.shape.email
```

kbventures committed:
Fix: parameter order for auth reset password

madcampos commented on api/src/routes/auth/validation.ts:
```diff
@@ -78,3 +79,26 @@ export type SignUpData = z.infer<typeof SignUpSchema>;
+
+export type ResetPasswordData = z.infer<typeof ResetPasswordSchema>;
+
+export const ResetPasswordValidTokenSchema = z.object({
+	token: z.uuid()
+});
+
+export type ResetPasswordValidTokenData = z.infer<typeof ResetPasswordValidTokenSchema>;
```
**question**: Why to we have this extra schema that is similar to the previous one? Can't we reuse parts of that one?

Can't we invert the logic so there is one schema with the token and then another one that extends the one with the token to also add a password?

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -234,6 +234,165 @@ authRoutes.openapi(
+authRoutes.openapi(
+	createRoute({
+		method: 'post',
+		path: '/valid-reset-pw-token',
```
```suggestion
		path: '/check-password-reset-token',
```

madcampos commented on api/src/routes/auth/index.ts:
```diff
@@ -234,6 +234,165 @@ authRoutes.openapi(
+authRoutes.openapi(
+	createRoute({
+		method: 'post',
+		path: '/valid-reset-pw-token',
```
**question**: Why do we have this endpoint?

kbventures replied:
In the event a user tries to utilize the reset-password page with an invalid token, they are redirected.

kbventures committed:
auth forgot password kv store ttl

madcampos reviewed (changes requested):
Please note:
- Some of the comments apply to global patterns, so they are about one place, and should be thought about how to apply them on the rest of the changes.
- I've avoided commenting multiple times the same thing to reduce the spam.
- There are still unresolved comments from previous reviews. Please resolve them.

kbventures committed:
Fix: Repalced clearTimeout with clearInterval to match setInterval

kbventures committed:
Refactaor validateToken useEffect in resetpassword component

kbventures committed:
Fix: css to rem and variable name

kbventures committed:
Update api/src/routes/auth/index.ts

Co-authored-by: Marco Campos <me@madcampos.dev>

kbventures committed:
Fix: package json reconsile name

kbventures committed:
Refactor: auth validate utilize existing shema

kbventures committed:
Refactored forgot-password tests

kbventures committed:
Removed content type since they are applied automaticaly for json hurl tests

kbventures committed:
Create variable for forgot password api tests

kbventures committed:
Retilized existing schema for ResetpasswordData zod Object

kbventures committed:
API Hurl Tests for forgot password and reset password auth routes

kbventures committed:
Fix: correct syntax for hurl variables

kbventures committed:
Fix/Refactor reset password hurl tests

kbventures committed:
Fix: minor api tests bugs

kbventures committed:
Minor api test fix

kbventures committed:
Merge branch 'dev' into feat/forgot-password

kbventures committed:
Remove testing token for reset password page

kbventures committed:
Fix: Missing file extension breaking build process

kbventures committed:
Fix: Missing file extension breaking build process

kbventures committed:
Removed forgot password states and replaced with derived states

kbventures committed:
Fix: spacing

kbventures committed:
Move function outside react function as an optimization

kbventures committed:
Move state management into handler and moved function outside reaction component for better resource utilization

kbventures committed:
Function expression for functions outside react component

kbventures committed:
Remove state and refactor to derived state

kbventures committed:
Merge Dev

kbventures committed:
Reset password todo notes

kbventures committed:
Heartbook hook added to sign in page

kbventures committed:
Hearthbeat sign in and log out home page

kbventures committed:
Fix: refactored console.error use

manvinderjit commented:
Why are local tests passing/skipped but remote ones are failing.

<img width="839" height="494" alt="image" src="https://github.com/user-attachments/assets/1e69b5df-e350-4c4f-8cdf-f180106f47b6" />

manvinderjit commented on wrangler.toml:
```diff
@@ -103,6 +113,11 @@ binding = "ActivationTokens"
 id = "987a04058cc34aad84b9f54c658f6a01"
 preview_id = "vms-production-activation-tokens"
 
+[[env.dev.kv_namespaces]]
```
shouldn't this be
```suggestion
[[env.prod.kv_namespaces]]
```

manvinderjit commented on src/pages/reset-password/reset-password.tsx:
```diff
@@ -0,0 +1,22 @@
+import { ResetPassword } from '../../components/ResetPassword/ResetPassword.tsx';
+import '../../index.css';
+
+const root = document.getElementById('root') as HTMLAnchorElement;
```
query: Why is the `HTMLAnchorElement` and not `HTMLElement`
```
const root = document.getElementById('root') as HTMLElement;
```

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,118 @@
+		const formData = new FormData(event.target as HTMLFormElement);
+		const emailValue = (formData.get('email') as string).trim().toLowerCase();
+
+		setIsSubmitted(true);
```
Why is this twice. Here and again on line 64.
I think this one should be removed.

manvinderjit commented on src/components/ForgotPassword/ForgotPassword.tsx:
```diff
@@ -0,0 +1,118 @@
+		const success = await requestPasswordRecovery(emailValue);
+
+		if (success) {
+			setIsSubmitted(true);
```
See above comment.

manvinderjit reviewed (dismissed):
Most of this looks good. Just a couple of questions and improvements.

kbventures committed:
Merge branch 'dev' into feat/forgot-password

kbventures commented:
> Why are local tests passing/skipped but remote ones are failing.
> 
> <img alt="image" width="839" height="494" src="https://private-user-images.githubusercontent.com/91133409/477750354-1e69b5df-e350-4c4f-8cdf-f180106f47b6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTUyMDM3NDEsIm5iZiI6MTc1NTIwMzQ0MSwicGF0aCI6Ii85MTEzMzQwOS80Nzc3NTAzNTQtMWU2OWI1ZGYtZTM1MC00YzRmLThjZGYtZjE4MDEwNmY0N2I2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA4MTQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwODE0VDIwMzA0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExYjMwZTcwMzY0M2M0ZGJiZTVmNzAxYjI4MGY4NzIxYWFkZWI2MTI1YjYwMWRiNjY0MjBhZjE3YWM2NmM0MGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oBQZLYm3lVNTyc6Ftn786adff8yL1d6CNbWOBqHQOUE">

@manvinderjit the unit tests are not include in the CICD pipeline at the moment.

kbventures committed:
Fix: auto complete error

kbventures committed:
Fix: wrangler spelling error and duplicate state set in forgotpassword page

kbventures committed:
Fix: incorect api test result
